apps/labs/posts/oss-internship-creating-narwhals-daft-plugin.md

```diff
@@ -0,0 +1,310 @@
+
+        `narwhals-<insert libraryname> = 'narwhals_<insert libraryname>'`
+
+2.  The two top-level functions in `__init__.py`:
+
+    - `is_native object`
+    - `from_native`
+    - the constant `NATIVE_PACKAGE` with the name of the package.
+
+3.  the 3 compliant classes: `CompliantLazyFrame`, `Expression` & `Namespace`; in our daft-specific use case, we have `DaftLazyFrame`,`DaftExpr` & `DaftNamespace`, other plugin developers, say for fictitious `grizzlies` library would then have `GrizzliesLazyFrame`,`GrizzliesExpr` & `GrizzliesNamespace`(4).
+
+## Issues we've considered when creating this solution
+
+## Finding the most lightweight approach to importing the framework of the plugin
+
+Within the plugin, it was important to find a way of detecting whether the framework the plugin is written for is present, without going through the slow step of loading the framework itself. This is so that we can avoid importing `daft` if the user is not actually using a daft dataframe. for example, if a user has installed the (not-yet-existent) plugins `narwhals-xarray`, `narwhals-daft`, `narwhals-grizzlies`, then it should be possible to detect which plugin to use for a given user input without having to import all of `xarray`, `daft`, and `grizzlies`. If you recall the structure of the top-level `__init__.py` file, we are indeed only importing daft once we're checking a dataframe.
+
+## Ensuring backwards-compatibility
+
+### Related materials: the case of plugins in scikit-learn
+
+It may be instructive to consider the issues faced by other packages when planning our own plugin.
+
+For scikit-learn, it is a known issue that even with minor releases, extensions currently break (true as of 01/09/25, see discussion [here](https://github.com/scikit-learn/scikit-learn/issues/31912)).
+
+There is currently a debate about how to fix this, with one proposal suggesting a more abstracted architecture where potentially breaking changes can be made safe by being placed between stable public-facing layers. On the other hand, this represents a quite significant rewrite of the package which may not sit well with users accustomed to more procedural scientific code. Scikit-learn is an established package with a long history and is widely used; this shows that questions around interfaces to other packages should be considered as early on as possible.
+
+### Narwhals
+
+Narwhals has a strong policy of [backwards compatibility](https://narwhals-dev.github.io/narwhals/backcompat/), mainly enforced through the use of a `stable` namespace. However, that promise does not extend to its internal Compliant protocols.
+
+Given that plugin authors use the Compliant Protocols, we need to ensure that these protocols are stable enough that users' plugins don't break with every minor Narwhals release
+
+**Here are some guidelines which we (Narwhals developers) aim to follow, to minimise the potential disruption on plugins authors.**
```

Comment:
Can you add a link to where this is documented for Narwhals?

Reply:
we haven't formalised this guidance in Narwhals yet, so I'd suggest rephrasing to highlight that this policy is something that Narwhals may consider adopting